chainerchem/dataset/preprocessors/weavenet_preprocessor.py

```diff
@@ -1,17 +1,166 @@
 from chainerchem.dataset.preprocessors.common import type_check_num_atoms
 from chainerchem.dataset.preprocessors.mol_preprocessor import MolPreprocessor
 
 
-MAX_DISTANCE = 7
-# TODO(Nakago): Review default size
-DEFAULT_NUM_MAX_ATOMS = 20  # 40
+ATOM = ['H', 'C', 'N', 'O', 'S', 'Cl', 'Br', 'F', 'P', 'I']
+MAX_DISTANCE = 2  # 7
+DEFAULT_NUM_MAX_ATOMS = 20  # 60  # paper
+
+
+# --- Atom feature extraction ---
+def construct_atom_type_vector(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom_type = len(ATOM)
+    n_atom = mol.GetNumAtoms()
+    atom_type_vector = numpy.zeros((num_max_atoms, n_atom_type),
+                                   dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        atom_idx = ATOM.index(a.GetSymbol())
+        atom_type_vector[i, atom_idx] = 1.0
+    return atom_type_vector
+
+
+def construct_formal_charge_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom = mol.GetNumAtoms()
+    formal_charge_vec = numpy.zeros((num_max_atoms, 1), dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        formal_charge_vec[i, 0] = a.GetFormalCharge()
+    return formal_charge_vec
+
+
+def construct_hybridization_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    n_atom = mol.GetNumAtoms()
+    hybridization_vec = numpy.zeros((num_max_atoms, 3), dtype=numpy.float32)
+    for i in range(n_atom):
+        a = mol.GetAtomWithIdx(i)
+        if a.GetHybridization() is None:
+            continue
+        hybridization_type = str(a.GetHybridization())
+        if hybridization_type == 'SP1':
+            hybridization_vec[i, 0] = 1.0
+        elif hybridization_type == 'SP2':
+            hybridization_vec[i, 1] = 1.0
+        elif hybridization_type == 'SP3':
+            hybridization_vec[i, 2] = 1.0
+    return hybridization_vec
+
+
+def construct_partial_charge_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    AllChem.ComputeGasteigerCharges(mol)
+    n = mol.GetNumAtoms()
+    partial_charge_vec = numpy.zeros((num_max_atoms, 1), dtype=numpy.float32)
+    for i in range(n):
+        a = mol.GetAtomWithIdx(i)
+        partial_charge_vec[i, 0] = a.GetProp("_GasteigerCharge")
+    return partial_charge_vec
+
+
+def construct_atom_ring_vec(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    nAtom = mol.GetNumAtoms()
+    rinfo = mol.GetRingInfo()
+    sssr = Chem.GetSymmSSSR(mol)
+    ring_feature = numpy.zeros((num_max_atoms, 6,), dtype=numpy.float32)
+    for ring in sssr:
+        ring = list(ring)
+        for i in range(nAtom):
+            if i in ring:
+                ring_size = len(ring)
+                if ring_size >= 3 and ring_size <= 8:
+                    ring_feature[i, ring_size - 3] = 1.0
+    return ring_feature
+
+
+def construct_hydrogen_bonding(mol, num_max_atoms=DEFAULT_NUM_MAX_ATOMS):
+    fdefName = os.path.join(RDConfig.RDDataDir, 'BaseFeatures.fdef')
+    factory = ChemicalFeatures.BuildFeatureFactory(fdefName)
```

Comment:
If `construct_hydrogen_couning` is called many time, we can make the factory once out side of this function to speed up. But you can do it later.

Reply:
Noted, it is also ok to raise an issue if it is important.